src/main/java/picard/sam/BamErrorMetric/BaseErrorCalculation.java

```diff
@@ -0,0 +1,266 @@
+import org.broadinstitute.barclay.argparser.CommandLineParser;
+
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Supplier;
+
+/**
+ * An interface and implementations for classes the look at a collection of bases in reads and computes various ErrorMetrics from them.
+ *
+ * @author Yossi Farjoun
+ */
+
+public class BaseErrorCalculation {
+
+    /**
+     * An interface that can take a collection of bases (provided as {@link htsjdk.samtools.util.SamLocusIterator.RecordAndOffset RecordAndOffset}
+     * and {@link SamLocusAndReferenceIterator.SAMLocusAndReference SAMLocusAndReference}) and generates a
+     * {@link ErrorMetrics.ErrorMetricBase} from them.
+     * <p>
+     * The Calculator has a suffix which will be used to generate the metric file suffixes.
+     */
+    public interface BaseCalculator {
+
+        /**
+         * The suffix that pertains to the implementation of aggregation
+         **/
+        String getSuffix();
+
+        /**
+         * Returns the metric generated by the observed loci
+         **/
+        ErrorMetrics.ErrorMetricBase getMetric();
+
+        /**
+         * the function by which new loci are "shown" to the calculator
+         **/
+        void addBase(final RecordAndOffset recordAndOffset,
+                     final SamLocusAndReferenceIterator.SAMLocusAndReference locusInfo);
+    }
+
+    /**
+     * An enum that is used to generate a {@link Supplier<BaseErrorCalculator>} from a string
+     * To use this given a String 'str':
+     * <p>
+     * Errors.valueOf(str).getErrorSupplier()
+     * <p>
+     * This is used in {@link CollectBamErrorMetrics} to convert an input argument to a {@link BaseErrorAggregation}.
+     */
+    enum Errors implements CommandLineParser.ClpEnum {
+        ERROR(SimpleErrorCalculator::new, "Collects the average error at the bases provided."),
+        OVERLAPPING_ERROR(OverlappingReadsErrorCalculator::new, "Only considers bases from the overlapping parts of reads from the same template. " +
+                "For those bases, it calculates the error that can be attributable to pre-sequencing, versus during-sequencing.");
+
+        private final Supplier<? extends BaseCalculator> errorSupplier;
+
+        Errors(Supplier<? extends BaseCalculator> errorSupplier, final String docString) {
+            this.errorSupplier = errorSupplier;
+            this.docString = docString;
+        }
+
+        public Supplier<? extends BaseCalculator> getErrorSupplier() {
+            return errorSupplier;
+        }
+
+        private final String docString;
+
+        @Override
+        public String getHelpDoc() {
+            return docString + " Suffix is: '" + errorSupplier.get().getSuffix() + "'.";
+        }
+    }
+
+    public abstract static class BaseErrorCalculator implements BaseCalculator {
+        long totalBases = 0;
+
+        /**
+         * the function by which new loci are "shown" to the calculator
```

Comment:
In general, you don't want to javadoc an overridden method. It's unnecessary as the base class javadoc is inherited automatically and, even worse, can lead to skew if the base class doc is changed.

An exception is if the override changes the behavior of a method in a way that a caller would need to know about.